File: SubArraySumEqualsK.cpp
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
// This problem is implemented using running sum approach.

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {

        if(nums.size()==0)
            return 0;
        unordered_map<int,int> map;
        int sum =0,count =0;
        map[0] = 1;

        for(int i=0;i<nums.size();i++)
        {
            sum = sum + nums[i];
            if(map.find(sum - k)!= map.end())
            {
                count = count + map[sum-k];
            }
            map[sum]++;
        }
        return count;

    }
};

File: subarray_sum_equals_k.py
"""Time complexity: O(n)
Space complexity: O(n)"""

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        n = len(nums)
        count=0
        sum=0
        sum_map = {0:1}
        for i in range(n):
            sum +=nums[i]
            if sum-k in sum_map:
                count+=sum_map[sum-k]
            sum_map[sum] = sum_map.get(sum, 0)+1
        return count